src/intralinks.rs

```diff
@@ -0,0 +1,933 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use syn::export::fmt::Display;
+use syn::export::{Debug, Formatter};
+use syn::Ident;
+use syn::Item;
+
+pub enum IntraLinkError {
+  IOError(std::io::Error),
+  ParseError(syn::Error),
+  NoModuleFile(String, PathBuf),
+}
+
+impl fmt::Display for IntraLinkError {
+  fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+    match self {
+      IntraLinkError::IOError(err) => write!(f, "IO error: {}", err),
+      IntraLinkError::ParseError(err) => write!(f, "NoModuleFile error: {}", err),
+      IntraLinkError::NoModuleFile(module, dir) => write!(
+        f,
+        "Unable to find module file for module {} in directory {:?}",
+        module, dir
+      ),
+    }
+  }
+}
+
+impl From<std::io::Error> for IntraLinkError {
+  fn from(err: std::io::Error) -> Self {
+    IntraLinkError::IOError(err)
+  }
+}
+
+impl From<syn::Error> for IntraLinkError {
+  fn from(err: syn::Error) -> Self {
+    IntraLinkError::ParseError(err)
+  }
+}
+
+fn file_ast<P: AsRef<Path>>(file: P) -> Result<syn::File, IntraLinkError> {
+  let mut file = File::open(file)?;
+  let mut src = String::new();
+
+  file.read_to_string(&mut src)?;
+
+  Ok(syn::parse_file(&src)?)
+}
+
+/// Determines the module filename, which can be `<module>.rs` or `<module>/mod.rs`.
+fn module_filename(dir: &Path, module: &Ident) -> Option<PathBuf> {
+  let mod_file = dir.join(format!("{}.rs", module));
+
+  if mod_file.is_file() {
+    return Some(mod_file);
+  }
+
+  let mod_file = dir.join(module.to_string()).join("mod.rs");
+
+  if mod_file.is_file() {
+    return Some(mod_file);
+  }
+
+  None
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub enum FQIdentifierAnchor {
+  Root,
+  Crate,
+}
+
+/// Fully qualified identifier.
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub struct FQIdentifier {
+  pub anchor: FQIdentifierAnchor,
+  pub path: Vec<String>,
+}
+
+impl FQIdentifier {
+  fn new(anchor: FQIdentifierAnchor) -> FQIdentifier {
+    FQIdentifier {
+      anchor,
+      path: Vec::new(),
+    }
+  }
+
+  fn from_string(s: &str) -> Option<FQIdentifier> {
+    let anchor;
+    let str;
+
+    if s == "::" {
+      return Some(FQIdentifier::new(FQIdentifierAnchor::Root));
+    } else if s.starts_with("::") {
+      anchor = FQIdentifierAnchor::Root;
+      str = &s[2..];
+    } else if s == "crate" {
+      return Some(FQIdentifier::new(FQIdentifierAnchor::Crate));
+    } else if s.starts_with("crate::") {
+      anchor = FQIdentifierAnchor::Crate;
+      str = &s[7..];
+    } else {
+      return None;
+    }
+
+    if str.is_empty() {
+      return None;
+    }
+
+    Some(FQIdentifier {
+      anchor,
+      path: str.split("::").map(|s| s.to_owned()).collect(),
```

Comment:
You can replace the lambda with:

```suggestion
      path: str.split("::").map(str::to_owned).collect(),
```